newlang/hckernel/mlir/lib/Dialect/PyIR/IR/PyIROps.cpp

```diff
@@ -28,6 +28,61 @@ void hc::py_ir::PyIRDialect::initialize() {
       >();
 }
 
+namespace {
+
+struct CleanupUnusedCaptures final
+    : public mlir::OpRewritePattern<hc::py_ir::PyFuncOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(hc::py_ir::PyFuncOp op,
+                  mlir::PatternRewriter &rewriter) const override {
+    auto capturesBlockArgs = op.getCaptureBlockArgs();
+    auto captureNames = op.getCaptureNames();
+
+    bool hasUnused = false;
+    for (auto &&arg : capturesBlockArgs) {
+      if (arg.use_empty()) {
+        hasUnused = true;
+        break;
+      }
+    }
+
+    if (!hasUnused)
+      return mlir::failure();
+
+    mlir::SmallVector<unsigned> indexes;
+    for (int i = captureNames.size() - 1; i >= 0; --i) {
+      auto capt = capturesBlockArgs[i];
+      if (capt.use_empty()) {
+        indexes.push_back(i);
+      }
+    }
+
+    auto eraseCaptureByIndex = [&](unsigned idx) {
+      int blockArgIdx = idx + op.getBlockArgs().size();
+      auto *entryBlock = op.getEntryBlock();
+      op.getCaptureArgsMutable().erase(idx);
+      auto captNames = mlir::SmallVector<mlir::Attribute>(
+          op.getCaptureNames().getAsRange<mlir::Attribute>());
+      captNames.erase(captNames.begin() + idx);
+      auto attr =
+          rewriter.getArrayAttr(mlir::ArrayRef<mlir::Attribute>(captNames));
+      op.setCaptureNamesAttr(attr);
+      entryBlock->eraseArgument(blockArgIdx);
+    };
+
+    for (auto &&idx : indexes) {
+      eraseCaptureByIndex(idx);
```

Comment:
You need to wrap inplace modifications in `modifyOpInPlace`

Reply:
done